### PR TITLE
BLOC-422 -- betting api

### PR DESCRIPTION
**Pull request by alexplevako**
https://scorum.atlassian.net/browse/BLOC-422

### CODE REVIEW

File: libraries/app/include/scorum/app/betting_api_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+#include <scorum/utils/range_adaptors.hpp>
+
+#include <scorum/utils/take_n_range_adaptor.hpp>
+
+namespace scorum {
+namespace app {
+
+#define BETTING_API_MAX_QUERY_LIMIT 1000
+
+template <typename T> typename T::iterator get_end_interator(T& collection, size_t limit)
```
Comment (Andrei-Masilevich) [inline libraries/app/include/scorum/app/betting_api_impl.hpp]: Where is it used?

File: libraries/app/include/scorum/app/betting_api_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        , _max_query_limit(max_query_limit)
+    {
+    }
+
+    std::vector<game_api_object> get_games(game_filter filter) const
+    {
+        // clang-format off
+        auto rng = _game_service.get_games()
+                | boost::adaptors::filtered([&](const auto& obj) {
+                    if (static_cast<uint8_t>(obj.status) & static_cast<uint8_t>(filter))
```
Comment (Andrei-Masilevich) [inline libraries/app/include/scorum/app/betting_api_impl.hpp]: What is the hack? Use == operator

File: tests/utests/utils/take_n_range_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+    BOOST_CHECK(!fst_foo.was_moved);
+    BOOST_CHECK(fst_foo.was_copied);
+}
+
+BOOST_AUTO_TEST_CASE(rvalue_underlying_range_move_test)
+{
+    auto rng = get_foo_vector() | utils::adaptors::take_n(1);
+    auto& fst_foo = *rng.begin();
+
```
Comment (Andrei-Masilevich) [inline tests/utests/utils/take_n_range_tests.cpp]: Add
    BOOST_CHECK(fst_foo.was_moved);
or use only one flags instead was_moved, was_copied

File: libraries/app/betting_api.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
+#include <scorum/app/betting_api.hpp>
+#include <scorum/app/betting_api_impl.hpp>
+
+namespace scorum {
+namespace app {
+
+betting_api::betting_api(const api_context& ctx)
+    : _impl(new betting_api_impl(*ctx.app.chain_database()))
```
Comment (Katkevich) [inline libraries/app/betting_api.cpp]: Use std::make_unique

File: libraries/app/include/scorum/app/betting_api_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+#include <boost/range/adaptor/filtered.hpp>
+
+#include <scorum/utils/range_adaptors.hpp>
+
+#include <scorum/utils/take_n_range_adaptor.hpp>
+
+namespace scorum {
+namespace app {
+
+#define BETTING_API_MAX_QUERY_LIMIT 1000
```
Comment (Katkevich) [inline libraries/app/include/scorum/app/betting_api_impl.hpp]: We have common approach for such a limit. See `get_api_config` function. It allows to configure limits via config

File: libraries/app/include/scorum/app/betting_api_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        , _max_query_limit(max_query_limit)
+    {
+    }
+
+    std::vector<game_api_object> get_games(game_filter filter) const
+    {
+        // clang-format off
+        auto rng = _game_service.get_games()
+                | boost::adaptors::filtered([&](const auto& obj) {
+                    if (static_cast<uint8_t>(obj.status) & static_cast<uint8_t>(filter))
```
Comment (Katkevich) [inline libraries/app/include/scorum/app/betting_api_impl.hpp]: Use just `return static_cast<uint8_t>(obj.status) & static_cast<uint8_t>(filter)`.

File: libraries/app/include/scorum/app/betting_api_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                        return false;
+                    }
+                })
+                | boost::adaptors::transformed([](const auto& obj) { return game_api_object(obj); });
+        // clang-format on
+
+        return { rng.begin(), rng.end() };
+    }
+
+    auto get_user_bets(bet_id_type from, int64_t limit) const
```
Comment (Katkevich) [inline libraries/app/include/scorum/app/betting_api_impl.hpp]: Use unsigned type for limit

**Pull request merged**